File: apps/frontend/src/lib/sentenceAlignment.ts
export type SentenceLanguage = 'en' | 'ja';

export interface SentenceSegment {
  key: string;
  pairKey: string | null;
  displayIndex: number;
  text: string;
  start: number;
  end: number;
}

export interface SentenceParagraph {
  key: string;
  sentences: SentenceSegment[];
}

export interface SentenceAlignment {
  englishParagraphs: SentenceParagraph[];
  japaneseParagraphs: SentenceParagraph[];
}

const paragraphBreakPattern = /((?:\r?\n)[\t ]*(?:\r?\n)+)/g;
const englishSentencePattern = /[^.!?]+[.!?]+["')\]]*|[^.!?]+$/g;
const japaneseSentencePattern = /[^。！？!?]+[。！？!?]+["'）】」』]*|[^。！？!?]+$/gu;
const protectedDot = '\u0000';
const closingPunctuation = new Set(['"', "'", '”', '’', ')', ']', '}', '』', '」']);
const sentenceStartAfterInitialismPattern = /^["')\]}”’]*\s+(?:A|An|The|I|We|You|He|She|It|They|This|That|These|Those|However|Therefore|Meanwhile|Instead|Finally|Next|Then)\b/u;

type SentenceSplitMode = 'legacy' | 'deterministic';
type SentenceSegmenter = {
  segment: (input: string) => Iterable<{ segment: string; index: number }>;
};

const getEnglishSentenceSegmenter = (): SentenceSegmenter | null => {
  const SegmenterCtor = (Intl as unknown as {
    Segmenter?: new (locale: string, options: { granularity: 'sentence' }) => SentenceSegmenter;
  }).Segmenter;
  return typeof SegmenterCtor === 'function' ? new SegmenterCtor('en', { granularity: 'sentence' }) : null;
};

const protectEnglishDots = (text: string): string => {
  let protectedText = text
    .replace(/(\d)\.(?=\d)/g, `$1${protectedDot}`);
  protectedText = protectedText.replace(
    /\b(?:[A-Za-z]\.){2,}/gi,
    (match, offset: number, source: string) => {
      const alwaysInternal = match.toLowerCase() === 'e.g.' || match.toLowerCase() === 'i.e.';
      const nextText = source.slice(offset + match.length);
      if (alwaysInternal || !sentenceStartAfterInitialismPattern.test(nextText)) {
        return match.replace(/\./g, protectedDot);
      }
      return `${match.slice(0, -1).replace(/\./g, protectedDot)}.`;
    },
  );
  protectedText = protectedText.replace(
    /\b(?:etc|vs)\./gi,
    (match, offset: number, source: string) => {
      const nextText = source.slice(offset + match.length);
      return sentenceStartAfterInitialismPattern.test(nextText)
        ? match
        : match.replace(/\./g, protectedDot);
    },
  );
  protectedText = protectedText.replace(
    /\b(?:Mr|Mrs|Ms|Dr|Prof|Sr|Jr|St|No|e\.g|i\.e)\./gi,
    (match) => match.replace(/\./g, protectedDot),
  );
  protectedText = protectedText.replace(
    /([A-Za-z])\.(?=[A-Za-z])/g,
    `$1${protectedDot}`,
  );
  return protectedText;
};

const trimSentenceSegment = (
  raw: string,
  absoluteStart: number,
): { text: string; start: number; end: number } | null => {
  const leading = raw.match(/^\s*/)?.[0].length ?? 0;
  const trailing = raw.match(/\s*$/)?.[0].length ?? 0;
  const start = absoluteStart + leading;
  const end = absoluteStart + raw.length - trailing;
  if (end <= start) return null;
  return { text: raw.slice(leading, raw.length - trailing), start, end };
};

export const splitTextSentences = (
  text: string,
  paragraphStart: number,
  language: SentenceLanguage,
): Array<{ text: string; start: number; end: number }> => {
  const protectedText = language === 'en' ? protectEnglishDots(text) : text;
  const terminators = language === 'ja'
    ? new Set(['。', '！', '？', '!', '?'])
    : new Set(['.', '!', '?']);
  const sentences: Array<{ text: string; start: number; end: number }> = [];
  let sentenceStart = 0;
  let index = 0;
  while (index < protectedText.length) {
    if (!terminators.has(protectedText[index])) {
      index += 1;
      continue;
    }
    let end = index + 1;
    while (end < protectedText.length && terminators.has(protectedText[end])) end += 1;
    while (end < protectedText.length && closingPunctuation.has(protectedText[end])) end += 1;
    if (language === 'ja' || end === protectedText.length || /\s/u.test(protectedText[end])) {
      const sentence = trimSentenceSegment(
        text.slice(sentenceStart, end),
        paragraphStart + sentenceStart,
      );
      if (sentence) sentences.push(sentence);
      sentenceStart = end;
    }
    index = end;
  }
  if (sentenceStart < text.length) {
    const sentence = trimSentenceSegment(
      text.slice(sentenceStart),
      paragraphStart + sentenceStart,
    );
    if (sentence) sentences.push(sentence);
  }
  if (sentences.length) return sentences;

  const fallback = text.trim();
  if (!fallback) return [];
  const start = paragraphStart + (text.match(/^\s*/)?.[0].length ?? 0);
  return [{ text: fallback, start, end: start + fallback.length }];
};

const splitLegacyTextSentences = (
  text: string,
  paragraphStart: number,
  language: SentenceLanguage,
): Array<{ text: string; start: number; end: number }> => {
  if (language === 'en') {
    const segmenter = getEnglishSentenceSegmenter();
    if (segmenter) {
      const segmented = Array.from(segmenter.segment(text))
        .map((segment) => trimSentenceSegment(segment.segment, paragraphStart + segment.index))
        .filter((sentence): sentence is { text: string; start: number; end: number } => Boolean(sentence));
      if (segmented.length) return segmented;
    }
  }

  const pattern = language === 'ja'
    ? new RegExp(japaneseSentencePattern.source, japaneseSentencePattern.flags)
    : new RegExp(englishSentencePattern.source, englishSentencePattern.flags);
  const sentences: Array<{ text: string; start: number; end: number }> = [];
  let match = pattern.exec(text);
  while (match) {
    const sentence = trimSentenceSegment(match[0], paragraphStart + match.index);
    if (sentence) sentences.push(sentence);
    match = pattern.exec(text);
  }
  if (sentences.length) return sentences;

  const fallback = text.trim();
  if (!fallback) return [];
  const start = paragraphStart + (text.match(/^\s*/)?.[0].length ?? 0);
  return [{ text: fallback, start, end: start + fallback.length }];
};

export const buildLanguageParagraphs = (
  value: string,
  language: SentenceLanguage,
  mode: SentenceSplitMode = 'deterministic',
): SentenceParagraph[] => {
  const chunks = value.split(paragraphBreakPattern);
  const paragraphs: SentenceParagraph[] = [];
  let cursor = 0;
  let sentenceCounter = 0;

  chunks.forEach((chunk) => {
    if (!chunk) return;
    if (new RegExp(paragraphBreakPattern.source).test(chunk)) {
      cursor += chunk.length;
      return;
    }

    const leading = chunk.match(/^\s*/)?.[0].length ?? 0;
    const trailing = chunk.match(/\s*$/)?.[0].length ?? 0;
    const start = cursor + leading;
    const end = cursor + chunk.length - trailing;
    if (end > start) {
      const paragraphIndex = paragraphs.length;
      const paragraphText = value.slice(start, end);
      const splitSentences = mode === 'legacy' ? splitLegacyTextSentences : splitTextSentences;
      const sentences = splitSentences(paragraphText, start, language).map((sentence, sentenceIndex) => {
        sentenceCounter += 1;
        return {
          key: `${language}-p${paragraphIndex}-s${sentenceIndex}`,
          pairKey: null,
          displayIndex: sentenceCounter,
          ...sentence,
        };
      });
      if (sentences.length) {
        paragraphs.push({
          key: `${language}-p${paragraphIndex}`,
          sentences,
        });
      }
    }
    cursor += chunk.length;
  });

  return paragraphs;
};

export const flattenSentenceParagraphs = (paragraphs: SentenceParagraph[]) => (
  paragraphs.flatMap((paragraph) => paragraph.sentences)
);

export const countSentencePairs = (alignment: SentenceAlignment): number => {
  const englishPairedCount = flattenSentenceParagraphs(alignment.englishParagraphs)
    .filter((sentence) => Boolean(sentence.pairKey)).length;
  const japanesePairedCount = flattenSentenceParagraphs(alignment.japaneseParagraphs)
    .filter((sentence) => Boolean(sentence.pairKey)).length;
  return Math.min(englishPairedCount, japanesePairedCount);
};

const regroupJapaneseParagraphs = (
  englishParagraphs: SentenceParagraph[],
  japaneseParagraphs: SentenceParagraph[],
): SentenceParagraph[] => {
  if (englishParagraphs.length <= 1 || japaneseParagraphs.length === englishParagraphs.length) {
    return japaneseParagraphs;
  }
  const englishCounts = englishParagraphs.map((paragraph) => paragraph.sentences.length);
  const englishSentenceTotal = englishCounts.reduce((sum, count) => sum + count, 0);
  const japaneseSentences = flattenSentenceParagraphs(japaneseParagraphs);
  if (japaneseSentences.length !== englishSentenceTotal) {
    return japaneseParagraphs;
  }

  let cursor = 0;
  return englishCounts.map((count, paragraphIndex) => {
    const sentences = japaneseSentences.slice(cursor, cursor + count).map((sentence, sentenceIndex) => ({
      ...sentence,
      key: `ja-p${paragraphIndex}-s${sentenceIndex}`,
    }));
    cursor += count;
    return {
      key: `ja-p${paragraphIndex}`,
      sentences,
    };
  });
};

const assignPairKeys = (
  englishParagraphs: SentenceParagraph[],
  japaneseParagraphs: SentenceParagraph[],
): SentenceAlignment => {
  const englishSentences = flattenSentenceParagraphs(englishParagraphs);
  const japaneseSentences = flattenSentenceParagraphs(japaneseParagraphs);
  const pairableCount = Math.min(englishSentences.length, japaneseSentences.length);
  const applyPairKeys = (paragraphs: SentenceParagraph[]) => paragraphs.map((paragraph) => ({
    ...paragraph,
    sentences: paragraph.sentences.map((sentence) => {
      const pairIndex = sentence.displayIndex - 1;
      return {
        ...sentence,
        pairKey: pairIndex < pairableCount ? `sentence-${pairIndex + 1}` : null,
      };
    }),
  }));

  return {
    englishParagraphs: applyPairKeys(englishParagraphs),
    japaneseParagraphs: applyPairKeys(japaneseParagraphs),
  };
};

export const buildSentenceAlignment = (
  bodyEn: string,
  bodyJa?: string | null,
  mode: SentenceSplitMode = 'legacy',
): SentenceAlignment => {
  const englishParagraphs = buildLanguageParagraphs(bodyEn, 'en', mode);
  const japaneseBaseParagraphs = bodyJa ? buildLanguageParagraphs(bodyJa, 'ja', mode) : [];
  const japaneseParagraphs = regroupJapaneseParagraphs(englishParagraphs, japaneseBaseParagraphs);
  return assignPairKeys(englishParagraphs, japaneseParagraphs);
};

export const createManualSentenceSegment = (
  key: string,
  pairKey: string | null,
  displayIndex: number,
  text: string,
): SentenceSegment => ({
  key,
  pairKey,
  displayIndex,
  text,
  start: 0,
  end: text.length,
});

File: apps/backend/backend/domain/quiz/sentence_alignment.py
from __future__ import annotations

import re
from dataclasses import dataclass


_PROTECTED_DOT = "\u0000"
_PARAGRAPH_BREAK = re.compile(r"\n\s*\n")
_ENGLISH_DOT_PATTERNS = (
    re.compile(r"(?<=\d)\.(?=\d)"),
    re.compile(r"\b(?:[A-Za-z]\.){2,}", re.IGNORECASE),
    re.compile(
        r"\b(?:Mr|Mrs|Ms|Dr|Prof|Sr|Jr|St|vs|etc|No|e\.g|i\.e)\.",
        re.IGNORECASE,
    ),
    re.compile(r"(?<=[A-Za-z])\.(?=[A-Za-z])"),
)
_CLOSING_PUNCTUATION = frozenset('"\'”’)]}』」')


@dataclass(frozen=True)
class TranslationAlignmentIssue:
    reason: str
    paragraph_index: int | None
    english_paragraph_count: int
    japanese_paragraph_count: int
    english_sentence_count: int | None = None
    japanese_sentence_count: int | None = None


def split_paragraphs(text: str) -> list[str]:
    normalized = str(text or "").replace("\r\n", "\n").replace("\r", "\n")
    return [part.strip() for part in _PARAGRAPH_BREAK.split(normalized) if part.strip()]


def _protect_english_dots(text: str) -> str:
    protected = text
    for pattern in _ENGLISH_DOT_PATTERNS:
        protected = pattern.sub(
            lambda match: match.group(0).replace(".", _PROTECTED_DOT),
            protected,
        )
    return protected


def _split_sentences(
    text: str,
    *,
    terminators: frozenset[str],
    require_whitespace_after: bool,
) -> list[str]:
    source = str(text or "")
    if not source.strip():
        return []
    protected = _protect_english_dots(source) if "." in terminators else source
    sentences: list[str] = []
    start = 0
    index = 0
    while index < len(protected):
        if protected[index] not in terminators:
            index += 1
            continue
        end = index + 1
        while end < len(protected) and protected[end] in terminators:
            end += 1
        while end < len(protected) and protected[end] in _CLOSING_PUNCTUATION:
            end += 1
        if (
            not require_whitespace_after
            or end == len(protected)
            or protected[end].isspace()
        ):
            sentence = source[start:end].strip()
            if sentence:
                sentences.append(sentence)
            start = end
        index = end
    trailing = source[start:].strip()
    if trailing:
        sentences.append(trailing)
    return sentences


def split_english_sentences(text: str) -> list[str]:
    return _split_sentences(
        text,
        terminators=frozenset(".!?"),
        require_whitespace_after=True,
    )


def split_japanese_sentences(text: str) -> list[str]:
    return _split_sentences(
        text,
        terminators=frozenset("。！？!?"),
        require_whitespace_after=False,
    )


def translation_alignment_issue(
    body_en: str,
    body_ja: str,
) -> TranslationAlignmentIssue | None:
    english_paragraphs = split_paragraphs(body_en)
    japanese_paragraphs = split_paragraphs(body_ja)
    english_paragraph_count = len(english_paragraphs)
    japanese_paragraph_count = len(japanese_paragraphs)
    if english_paragraph_count != japanese_paragraph_count:
        return TranslationAlignmentIssue(
            reason="paragraph_count_mismatch",
            paragraph_index=None,
            english_paragraph_count=english_paragraph_count,
            japanese_paragraph_count=japanese_paragraph_count,
        )
    for paragraph_index, (english, japanese) in enumerate(
        zip(english_paragraphs, japanese_paragraphs, strict=True),
        start=1,
    ):
        english_sentence_count = len(split_english_sentences(english))
        japanese_sentence_count = len(split_japanese_sentences(japanese))
        if english_sentence_count != japanese_sentence_count:
            return TranslationAlignmentIssue(
                reason="sentence_count_mismatch",
                paragraph_index=paragraph_index,
                english_paragraph_count=english_paragraph_count,
                japanese_paragraph_count=japanese_paragraph_count,
                english_sentence_count=english_sentence_count,
                japanese_sentence_count=japanese_sentence_count,
            )
    return None


__all__ = [
    "TranslationAlignmentIssue",
    "split_english_sentences",
    "split_japanese_sentences",
    "split_paragraphs",
    "translation_alignment_issue",
]
